cmd/aries-agent-lambda/storage/dynamodb/dynamodb.go

```diff
@@ -0,0 +1,562 @@
+		return nil, err
+	}
+
+	if result.Item == nil {
+		return nil, spi.ErrDataNotFound
+	}
+
+	var item dbEntry
+	if err := dynamodbattribute.UnmarshalMap(result.Item, &item); err != nil {
+		fmt.Printf("GetTags error: %s\n", err)
+		return nil, err
+	}
+
+	return unmarshalTags(item.Tags), nil
+}
+
+func marshalTags(tags []spi.Tag) (string, error) {
+	var sb strings.Builder
+	for _, tag := range tags {
+		if strings.Contains(tag.Name, ":") || strings.Contains(tag.Name, ",") {
+			return "", fmt.Errorf(invalidTagName, tag.Name)
+		}
+
+		if strings.Contains(tag.Value, ":") || strings.Contains(tag.Name, ",") {
+			return "", fmt.Errorf(invalidTagValue, tag.Value)
+		}
+
+		sb.WriteString(fmt.Sprintf("%s:%s,", tag.Name, tag.Value))
+	}
+
+	return "[" + strings.TrimRight(sb.String(), ",") + "]", nil
+}
+
+func unmarshalTags(tags string) []spi.Tag {
+	parsedTags := make([]spi.Tag, 1)
+	tagValueStr := strings.Split(strings.Trim(tags, "[]"), ",")
+	for _, tvStr := range tagValueStr {
+		tv := strings.Split(tvStr, ":")
+		parsedTags = append(parsedTags, spi.Tag{
+			Name: tv[0],
+			Value: tv[1],
+		})
+	}
+
+	return parsedTags
+}
+
+// GetBulk fetches the values associated with the given keys.
+// If no data exists under a given key, then a nil []byte is returned for that value. It is not considered an error.
+// If any of the given keys are empty, then an error will be returned.
+func (m *dynamodbTable) GetBulk(keys ...string) ([][]byte, error) {
+	if len(keys) == 0 {
+		return nil, errors.New("keys slice must contain at least one key")
+	}
+
+	fmt.Printf("Bulk getting keys %s\n", keys)
+
+	for _, key := range keys {
+		if key == "" {
+			return nil, errEmptyKey
+		}
+	}
+
+	values := make([][]byte, len(keys))
+
+	for i, key := range keys {
+		var err error
+		values[i], err = m.Get(key)
+		if err != nil {
+			fmt.Printf("GetBulk error: %s\n", err)
+			return values, err
+		}
+	}
+
+	return values, nil
+}
+
+// Query returns all data that satisfies the expression. Expression format: TagName:TagValue.
+// If TagValue is not provided, then all data associated with the TagName will be returned.
+// For now, expression can only be a single tag Name + Value pair.
+// None of the current query options are supported
+// spi.WithPageSize will simply be ignored since it only relates to performance and not the actual end result.
+// spi.WithInitialPageNum and spi.WithSortOrder will result in an error being returned since those options do
+// affect the results that the Iterator returns.
+func (m *dynamodbTable) Query(expression string, options ...spi.QueryOption) (spi.Iterator, error) {
+	err := checkForUnsupportedQueryOptions(options)
+	if err != nil {
+		fmt.Printf("Query error: %s\n", err)
+		return nil, err
+	}
+
+	// TODO
+
+	fmt.Printf("Querying %s\n", expression)
+
+	if expression == "" {
+		return nil, errInvalidQueryExpressionFormat
+	}
+
+	expr, err := dbexpr.NewBuilder().WithFilter(dbexpr.Name("Tags").Contains(expression)).Build()
```

Comment:
Perhaps imperfect (would need tweaks not to get substrings, like `id:bet` maching `id:better`, but it was performing nicely in our tests).